### ISSUE TITLE
L029 flags datatype identifiers incorrectly in tsql

### ISSUE THREAD

**Issue by alanmcruickshank** [sqlfluff/sqlfluff#2958]
### Search before asking

- [X] I searched the [issues](https://github.com/sqlfluff/sqlfluff/issues) and found no similar issues.

### What Happened

This query:

```sql
select cast(a as date) as b from c
```

...presents this error (when running in `tsql` but not `ansi` or `snowflake`):
```
L:   1 | P:  18 | L029 | Keywords should not be used as identifiers.
```

...when the `unquoted_identifiers_policy` is set to `all`.

This is because in the parse tree `cast(a as date)` is interpreted as:

```
[L:  1, P:  8]      |                        function:
[L:  1, P:  8]      |                            function_name:
[L:  1, P:  8]      |                                keyword:                      'cast'
[L:  1, P: 12]      |                            bracketed:
[L:  1, P: 12]      |                                start_bracket:                '('
[L:  1, P: 13]      |                                [META] indent:
[L:  1, P: 13]      |                                expression:
[L:  1, P: 13]      |                                    column_reference:
[L:  1, P: 13]      |                                        identifier:           'a'
[L:  1, P: 14]      |                                whitespace:                   ' '
[L:  1, P: 15]      |                                keyword:                      'as'
[L:  1, P: 17]      |                                whitespace:                   ' '
[L:  1, P: 18]      |                                data_type:
[L:  1, P: 18]      |                                    identifier:               'date'
[L:  1, P: 22]      |                                [META] dedent:
[L:  1, P: 22]      |                                end_bracket:                  ')'
```

Where `date` is correctly identified as a `data_type`, but within that still contains an `identifier`. In `ansi` and `snowflake` (and I suspect many others), the parse tree is uses `data_type_identifier` rather than just `identifier` and so does not trigger L029.

### Expected Behaviour

The behaviour should be consistent between all dialects.

### Observed Behaviour

`tsql` dialect fails where others pass.

### How to reproduce

Query is above. Config file below. Run `lint`.

### Dialect

TSQL

### Version

sqlfluff 0.11.1

### Configuration

```
[sqlfluff]
dialect = tsql

[sqlfluff:rules:L029]
unquoted_identifiers_policy = all
```

### Are you willing to work on and submit a PR to address the issue?

- [ ] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/sqlfluff/sqlfluff/blob/main/CODE_OF_CONDUCT.md)
